File: 206. reverse linked list.py
"""
Reverse a linked List
"""


class Solution:
    # iterative way
    def reverseList1(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        if not head or not head.next:
            return head

        current = head
        last = None
        while current:
            temp = current.next
            current.next = last
            last = current
            current = temp

        return last

    # recursive way
    def reverseList2(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        if not head:
            return

        last = None
        current = head

        ans = self.reverse(current, last)
        return ans

    def reverse(self, current, last):
        if not current:
            return last
        else:
            temp = current.next
            current.next = last
            last = current
            current = temp

            return self.reverse(current, last)


    # recursive way, but go to the final node first, then get back
    def reverseList(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        return self.helper(head, None)

    def helper(self, current, last):
        if current:
            res = self.helper(current.next, current)
            current.next = last # change next
        else:
            res = last # only assign value for res here
        return res

File: 206. Reverse Linked List.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    // iterative way
    ListNode* reverseList1(ListNode* head) {
        if(!head) return head;

        ListNode *prev = NULL, *current=NULL, *tmp=NULL;
        current = head;
        while(current != NULL) {
            tmp = current -> next;
            current -> next = prev;
            prev = current;
            current = tmp;
        }
        return prev;
    }


    // recursive way
    ListNode* helper(ListNode* cur, ListNode* prev) {
        if(!cur){
            return prev;
        }

        ListNode* rearHead = helper(cur->next, cur);
        cur -> next = prev;
        return rearHead;
    }
    ListNode* reverseList(ListNode* head) {
        if(!head) return head;
        return helper(head, NULL);
    }

};
